### DISCUSSION TITLE
About LIFNode

### DISCUSSION THREAD

**Discussion by Redamancy-520** [fangwei123456/SpikingFlow#213] [Q&A] [answered]
您好，我在学习您实现LIF神经元的代码。我想问一下如何控制神经元的输入电压X。例如我想在不同时刻输入不同的电压观察膜电位变化，只能手动更新X吗？

**Comment by fangwei123456** [comment, marked as answer]
是的，提前生成好不同时刻的x[t]，然后再输入

**Reply by Redamancy-520** [reply]
也就说是lif的输入x意味着在这一段时间内的持续输入？

**Reply by fangwei123456** [reply]
x[t]意味着t时间步的输入，参考神经元的教程。

**Reply by fangwei123456** [reply]
实际上可以认为是在一个时间步的持续输入。不过，由于时间被离散化了，“持续”一词因而也失去了意义。

**Reply by Toadally-Awesome** [reply]
> 是的，提前生成好不同时刻的x[t]，然后再输入

您好！关于这个从Relu的特征出发，寻找到脉冲发射频率与输入的相关性，所作出的Neuron设计有一点疑问。

在有一些文章中提到在ANN与SNN转换里采用某种首脉冲编码（TTFS），即只考虑每个神经元从接收到刺激开始放出的第一个脉冲，同时引入了一个“权重斜率”或“膜电位斜率”的参数μ，从而让上一层的某个神经元首脉冲发射时刻开始，对下一层与之相连的Neurons膜电位产生持续的影响（μ改变膜电位的增速）。我猜想这和教程中提到的IF Neuron可以作为Relu在时间上的无偏估计有类似的含义。

因此想请教一下，**Spiking Jelly的整体设计中，以及像是ANN2SNN转换以及Fashion-MNIST这两个文档教程中，采用的还是传统Spike-train的方式来编码脉冲序列的吗，即Neuron可以在人为规定的time steps中不限次数地发射脉冲到下一层？**

我目前在研究SNN的硬件设计，所以在软件模型和代码的理解使用上不是非常熟练，根据文档内“传播方式”的内容，并且如果Spiking Jelly的前向传播模式是允许同一个Neuron反复发射脉冲的，**那前向传播是否可以理解成：**

对某一层输入了一个为[Time, C, W, H]这样一个四维数据：
第零步：  初始化模型内膜电位，和中间层的脉冲数据。

第一步：  for t in range(0, Time) ：
                     作 [C, W, H]与权重的卷积，  单次获得[C', W', H']的特征图(Δ膜电位)，
                总计获得#Time个特征图。

第二步：  for i in range(0, Time)：
                     膜电位 += 特征图[i], 
                     根据膜电位，发射脉冲[C', W', H']的脉冲表
                总计获得#Time个脉冲表，叠起来即构成[Time, C', W', H']。

第三步：将[Time, C', W', H']作为下一层输入（跳转第一步）。

而不是将t-1时刻的输入，对整个模型内每一层的膜电位和脉冲激发情况，都计算完之后，再计算t时刻输入对每一层的影响。

**Comment by fangwei123456** [comment]
> 即Neuron可以在人为规定的time steps中不限次数地发射脉冲到下一层

是的，没有限制脉冲发放次数。但是很容易继承并新定义一个神经元，限制只能发放单次脉冲。在讨论区有一个类似的帖子。

> 那前向传播是否可以理解成：
> ...
> 而不是将t-1时刻的输入，对整个模型内每一层的膜电位和脉冲激发情况，都计算完之后，再计算t时刻输入对每一层的影响。

逐层传播模式是这样的。逐步传播则是单个时刻的输入，经过整个网络并计算，然后再输入下一个时刻的数据。

**Reply by Toadally-Awesome** [reply]
感谢回答！
还有一个小问题，可以监视模型内，每个time step时刻，任一层神经元的即时膜电位吗？

**Reply by fangwei123456** [reply]
用钩子函数可以轻松实现，另外SJ框架里面有一个监视器模块，提供了类似的功能。

**Reply by Toadally-Awesome** [reply]
钩子函数可以直接读取Clock_Driven中Neuron这个类的内部.v膜电位数据吗？
还是说需要自行通过设置膜电位初值为“0”，从t = 0开始，累加输入电位增量，并减去脉冲mask * Vthreshold（Vreset = none时）。

**Reply by fangwei123456** [reply]
钩子函数可以获取模块的任何内部变量，所以v是可以获取的。v在初始化时是0。
